Feature nondet string initialization [depends-on: #3750]

hannes-steffenhagen-diffblue opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

hannes-steffenhagen-diffblue committed:
Add support for dynamic pointer to array initialisation

hannes-steffenhagen-diffblue committed:
Add support for nondet initialisation of dynamic C strings

tautschnig commented on src/ansi-c/ansi_c_language.h:
```diff
@@ -22,13 +22,29 @@ Author: Daniel Kroening, kroening@kroening.com
   " --min-null-tree-depth N      minimum level at which a pointer can first be\n" /* NOLINT(*) */\
-  "                              NULL in a recursively nondet initialized struct\n" /* NOLINT(*) */
+  "                              NULL in a recursively nondet initialized struct\n" /* NOLINT(*) */\
+  " --pointers-to-treat-as-array <identifier,...>  Comma separated list of\n" \
```
Nit pick: we seem to start help messages with a lowercase letter.

tautschnig commented on src/ansi-c/ansi_c_language.h:
```diff
@@ -22,13 +22,29 @@ Author: Daniel Kroening, kroening@kroening.com
+  "                               identifiers that should be initialized as arrays\n" /* NOLINT(*) */ \
+  " --associated-array-sizes <identifier:identifier...>\n" \
+  "                               comma separated list of colon separated pairs\n" /* NOLINT(*) */ \
+  "                               of identifiers; The first should be the name\n" /* NOLINT(*) */ \
```
"The first" ... what?

tautschnig commented on src/ansi-c/c_nondet_symbol_factory.cpp:
```diff
@@ -13,16 +13,21 @@ Author: Diffblue Ltd.
 #include <goto-programs/goto_functions.h>
+#include <util/array_name.h>
+#include <util/optional_utils.h>
+#include <util/pointer_offset_size.h>
```
Place alongside the other includes from `util`

tautschnig commented on src/ansi-c/c_nondet_symbol_factory.cpp:
```diff
@@ -39,12 +45,14 @@ class symbol_factoryt
     const source_locationt &loc,
-    const c_object_factory_parameterst &object_factory_params)
+    const c_object_factory_parameterst &object_factory_params,
+    std::map<irep_idt, irep_idt> &deferred_array_sizes)
```
Where is this passed in as a non-empty map?

hannes-steffenhagen-diffblue replied:
This is kept as state between calls to `c_nondet_symbol_factory` here: https://github.com/diffblue/cbmc/pull/3572/commits/25be3699b8d0697498e4f4b95915ea4e18ada03b#diff-ffd0a317958e24d1f65f21ed40f67c99L39

tautschnig commented on src/ansi-c/ansi_c_language.h:
```diff
@@ -26,6 +26,7 @@ Author: Daniel Kroening, kroening@kroening.com
   "(pointers-to-treat-as-array):" \
   "(associated-array-sizes):" \
   "(max-dynamic-array-size):" \
+  "(pointers-to-treat-as-string):"
```
Missing help

hannes-steffenhagen-diffblue commented on src/ansi-c/c_nondet_symbol_factory.cpp:
```diff
@@ -53,24 +61,95 @@ class symbol_factoryt
+
+  /// Return the name of a variable holding an array size if one is associated
+  /// with the given symbol name
+  optionalt<dstringt> get_deferred_size(irep_idt symbol_name) const;
```
should be an irep_idt, damn you clion

tautschnig commented:
As a note while this is still "work in progress": I found it quite difficult to review. If there is any way to break things up into smaller commits, or maybe just into smaller hunks by refactoring functions that would be much appreciated.

danpoe commented on src/ansi-c/ansi_c_language.h:
```diff
@@ -22,13 +22,30 @@ Author: Daniel Kroening, kroening@kroening.com
   "(max-nondet-tree-depth):" \
-  "(min-null-tree-depth):"
+  "(min-null-tree-depth):" \
+  "(pointers-to-treat-as-array):" \
```
`pointers-to-treat-as-array` -> `pointers-to-treat-as-arrays` (and throughout the PR)

danpoe commented on src/ansi-c/ansi_c_language.h:
```diff
@@ -22,13 +22,30 @@ Author: Daniel Kroening, kroening@kroening.com
+  "(pointers-to-treat-as-array):" \
+  "(associated-array-sizes):" \
+  "(max-dynamic-array-size):" \
+  "(pointers-to-treat-as-string):"
```
`pointers-to-treat-as-string` -> `pointers-to-treat-as-strings` (and throughout the PR)

danpoe commented on src/ansi-c/c_nondet_symbol_factory.cpp:
```diff
@@ -13,16 +13,21 @@ Author: Diffblue Ltd.
 
 #include <ansi-c/c_object_factory_parameters.h>
 
+#include <functional>
```
Separate from the rest of the includes

danpoe commented on src/ansi-c/c_nondet_symbol_factory.cpp:
```diff
@@ -53,24 +61,95 @@ class symbol_factoryt
+  /// \param prefix: This forms the first part of the parameter,
+  ///   for debugging purposes must be a valid identifier prefix
+  /// \return A reference to the newly created symbol table entry
+  symbolt &new_tmp_symbol(const typet &type, const std::string &prefix);
```
This can now be replaced by `allocate_objectst::allocate_automatic_local_object()`

danpoe commented on src/ansi-c/c_nondet_symbol_factory.cpp:
```diff
@@ -53,24 +61,95 @@ class symbol_factoryt
+  /// variable yet (i.e. it is not yet in the symbol table and doesn't have
+  /// initialisation code generated for it yet). If that's the case we remember
+  /// that we have to set it to the right size later with this method.
+  /// \param associated_size_name: The of variable that should be set to
```
`The of variable...` -> `The name of the variable...`

danpoe commented on src/ansi-c/c_nondet_symbol_factory.cpp:
```diff
@@ -252,6 +455,270 @@ void symbol_factoryt::gen_nondet_array_init(
+    array_member_init_body,
+    dereference_exprt{
+      plus_exprt{array, array_index_symbol.symbol_expr(), array.type()}},
+    depth,
```
I think we need to pass in `depth + 1` here since we dereference the expression (since we pass in a `dereference_exprt`).

danpoe commented on src/ansi-c/c_nondet_symbol_factory.cpp:
```diff
@@ -228,11 +350,92 @@ void symbol_factoryt::gen_nondet_init(
+      if(associated_array_size.has_value())
+      {
+        assign.rhs() = typecast_exprt{
+          symbol_table.lookup_ref(associated_array_size.value()).symbol_expr(),
```
Hm, does this work? I thought `lookup_ref()` requires the full name of a symbol whereas `associated_array_size` is the base name I think.

hannes-steffenhagen-diffblue replied:
Ah, this does work but is confusingly worded. The `associated_array_size` here is actually the _deferred_ array size symbol which we dynamically created earlier, and we do have the full name for that. I'll see if I can make this clearer

hannes-steffenhagen-diffblue commented:
@tautschnig I'll see if I can restructure the commits to make the changes clearer. Meanwhile the first and second commit do implement related but separate features so it probably makes sense to keep them as separate PRs, which would also help a bit I hope

tautschnig commented:
@hannes-steffenhagen-diffblue Anything that can be done to make this a bit smaller would really help reviewing. If it's not possible we'll have to bite the bullet, so please just send a ping when this is in the best possible state for a review.

hannes-steffenhagen-diffblue commented:
@tautschnig I've extracted the larger part (that which is not string specific) into a separate PR at https://github.com/diffblue/cbmc/pull/3750. In there I've also split up that feature into a couple of mostly independent commits, which should make it a bit easier to review (although it is still not a small amount of code).

If that goes through I'll rebase the string specific bits and put them here.

hannes-steffenhagen-diffblue commented:
Closed as we decided to do this another way

tautschnig commented:
@hannes-steffenhagen-diffblue I'm a bit confused, as #3750 appears to have been closed as well?
